src/lib/gadgets/elliptic-curve.ts

```diff
@@ -0,0 +1,556 @@
+
+  // constant case
+  if (Point.isConstant(p1) && Point.isConstant(p2)) {
+    let p3 = affineAdd(Point.toBigint(p1), Point.toBigint(p2), f);
+    return Point.from(p3);
+  }
+
+  // witness and range-check slope, x3, y3
+  let witnesses = exists(9, () => {
+    let [x1_, x2_, y1_, y2_] = Field3.toBigints(x1, x2, y1, y2);
+    let denom = inverse(mod(x1_ - x2_, f), f);
+
+    let m = denom !== undefined ? mod((y1_ - y2_) * denom, f) : 0n;
+    let m2 = mod(m * m, f);
+    let x3 = mod(m2 - x1_ - x2_, f);
+    let y3 = mod(m * (x1_ - x3) - y1_, f);
+
+    return [...split(m), ...split(x3), ...split(y3)];
+  });
+  let [m0, m1, m2, x30, x31, x32, y30, y31, y32] = witnesses;
+  let m: Field3 = [m0, m1, m2];
+  let x3: Field3 = [x30, x31, x32];
+  let y3: Field3 = [y30, y31, y32];
+  ForeignField.assertAlmostFieldElements([m, x3, y3], f);
+
+  // (x1 - x2)*m = y1 - y2
+  let deltaX = ForeignField.Sum(x1).sub(x2);
+  let deltaY = ForeignField.Sum(y1).sub(y2);
+  ForeignField.assertMul(deltaX, m, deltaY, f);
+
+  // m^2 = x1 + x2 + x3
+  let xSum = ForeignField.Sum(x1).add(x2).add(x3);
+  ForeignField.assertMul(m, m, xSum, f);
+
+  // (x1 - x3)*m = y1 + y3
+  let deltaX1X3 = ForeignField.Sum(x1).sub(x3);
+  let ySum = ForeignField.Sum(y1).add(y3);
+  ForeignField.assertMul(deltaX1X3, m, ySum, f);
+
+  return { x: x3, y: y3 };
+}
+
+function double(p1: Point, f: bigint) {
+  let { x: x1, y: y1 } = p1;
+
+  // constant case
+  if (Point.isConstant(p1)) {
+    let p3 = affineDouble(Point.toBigint(p1), f);
+    return Point.from(p3);
+  }
+
+  // witness and range-check slope, x3, y3
+  let witnesses = exists(9, () => {
+    let [x1_, y1_] = Field3.toBigints(x1, y1);
+    let denom = inverse(mod(2n * y1_, f), f);
+
+    let m = denom !== undefined ? mod(3n * mod(x1_ ** 2n, f) * denom, f) : 0n;
+    let m2 = mod(m * m, f);
+    let x3 = mod(m2 - 2n * x1_, f);
+    let y3 = mod(m * (x1_ - x3) - y1_, f);
+
+    return [...split(m), ...split(x3), ...split(y3)];
+  });
+  let [m0, m1, m2, x30, x31, x32, y30, y31, y32] = witnesses;
+  let m: Field3 = [m0, m1, m2];
+  let x3: Field3 = [x30, x31, x32];
+  let y3: Field3 = [y30, y31, y32];
+  ForeignField.assertAlmostFieldElements([m, x3, y3], f);
+
+  // x1^2 = x1x1
+  let x1x1 = ForeignField.mul(x1, x1, f);
+
+  // 2*y1*m = 3*x1x1
+  // TODO this assumes the curve has a == 0
+  let y1Times2 = ForeignField.Sum(y1).add(y1);
+  let x1x1Times3 = ForeignField.Sum(x1x1).add(x1x1).add(x1x1);
+  ForeignField.assertMul(y1Times2, m, x1x1Times3, f);
+
+  // m^2 = 2*x1 + x3
+  let xSum = ForeignField.Sum(x1).add(x1).add(x3);
+  ForeignField.assertMul(m, m, xSum, f);
+
+  // (x1 - x3)*m = y1 + y3
+  let deltaX1X3 = ForeignField.Sum(x1).sub(x3);
+  let ySum = ForeignField.Sum(y1).add(y3);
+  ForeignField.assertMul(deltaX1X3, m, ySum, f);
+
+  return { x: x3, y: y3 };
+}
+
+function verifyEcdsa(
+  Curve: CurveAffine,
+  signature: Ecdsa.Signature,
+  msgHash: Field3,
+  publicKey: Point,
+  config: {
```

Comment:
Can we document the config a bit more and mention why we chose these default values?